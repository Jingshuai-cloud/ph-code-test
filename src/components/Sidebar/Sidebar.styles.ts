import styled from "styled-components";

export const NavWrapper = styled.nav`
  background: #f4f2f2;
  display: block;
  top: 0;
  bottom: 0;
  left: 0;
  width: 215px;
  border-right: 1px solid lightgrey;
  border-radius: 20px;
  position: absolute;
  left: 20px;
  top: 120px;
`;
