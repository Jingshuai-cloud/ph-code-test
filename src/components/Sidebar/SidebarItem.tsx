import React from "react";
import { ItemList, Item, ItemAnchor, ItemLabel } from "./Sidebar.styles";

export type MenuObject = {
  coffee: string;
  path: string;
  subMenu: Array<MenuObject>;
};

type Props = {
  menu: MenuObject;
  handleMenuSelection: any;
  selectedList: Array<string>;
  level: number;
};

const SidebarItem: React.FC<Props> = ({
  menu,
  handleMenuSelection,
  selectedList,
  level,
}) => {
  const { coffee, path, subMenu } = menu;
  return (
    <>
      {subMenu.length > 0 ? (
        <Item>
          <ItemLabel onMouseEnter={() => handleMenuSelection(coffee, level)}>
            {coffee}
          </ItemLabel>
          {selectedList[level] === coffee && (
            <ItemList>
              {subMenu.map((menu: MenuObject, i) => {
                const nextLevel = level + 1;
                return (
                  <SidebarItem
                    menu={menu}
                    handleMenuSelection={handleMenuSelection}
                    selectedList={selectedList}
                    level={nextLevel}
                    key={`child-${menu.coffee}-${i}`}
                  />
                );
              })}
            </ItemList>
          )}
        </Item>
      ) : (
        <Item>
          <ItemAnchor href={path}>{menu.coffee}</ItemAnchor>
        </Item>
      )}
    </>
  );
};

export default SidebarItem;
