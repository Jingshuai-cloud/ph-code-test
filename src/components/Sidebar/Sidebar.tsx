import React, { useState } from "react";
import { NavWrapper } from "./Sidebar.styles";
import SidebarItem from "./components/SidebarItem";
import { MenuObject } from "./components/SidebarItem";

type Props = {
  menus: Array<MenuObject>;
};

const Sidebar: React.FC<Props> = ({ menus }) => {
  const [selectedList, setSelectedList] = useState<string[]>([]);

  const handleMenuSelection = (subMenu: string, level: number) => {
    setSelectedList((selectedList: string[]) => {
      const newSelectedList: string[] = [...selectedList];

      newSelectedList.length = level;
      if (subMenu !== "") {
        newSelectedList[level] = subMenu;
      }
      return newSelectedList;
    });

    console.log(selectedList);
  };

  return (
    <NavWrapper>
      <ul onMouseLeave={() => setSelectedList([])}>
        {menus.map((menu: MenuObject, i) => {
          return (
            <SidebarItem
              menu={menu}
              key={`child-${menu.coffee}-${i}`}
              handleMenuSelection={handleMenuSelection}
              selectedList={selectedList}
              level={0}
            />
          );
        })}
      </ul>
    </NavWrapper>
  );
};

export default Sidebar;
