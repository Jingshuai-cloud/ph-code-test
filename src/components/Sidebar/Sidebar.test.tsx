import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Sidebar from "./Sidebar";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

const menus = [
  { coffee: "AFFOGATO", path: "/layer1", subMenu: [] },
  {
    coffee: "ESPRESSO",
    path: "/",
    subMenu: [
      { coffee: "ESPRESSO", path: "/layer2", subMenu: [] },
      { coffee: "DOUBLE ESPRESSO", path: "/layer2", subMenu: [] },
      {
        coffee: "MORE",
        path: "/",
        subMenu: [
          { coffee: "ESPRESSO CON PANNA", path: "/layer3", subMenu: [] },
          { coffee: "ESPRESSO MACCHIATO", path: "/layer3", subMenu: [] },
          {
            coffee: "MORE",
            path: "/",
            subMenu: [
              { coffee: "ICED LATTE", path: "/layer2", subMenu: [] },
              { coffee: "ICED MOCHA", path: "/layer2", subMenu: [] },
            ],
          },
        ],
      },
    ],
  },
];

describe("Sidebar", () => {
  test("renders Sidebar component", () => {
    render(<Sidebar menus={menus} />);

    expect(screen.getByRole("list")).toBeInTheDocument();
    const ESPRESSO = screen.getByText("ESPRESSO");
    expect(ESPRESSO).toBeInTheDocument();
  });

  it("should update selectedList on mouse enter", () => {
    const { container } = render(<Sidebar menus={menus} />);
    const selectedList = container.querySelector(".selectedList");
    //const ex = screen.getByRole("link");
    const li = screen.getByText("ESPRESSO");
    //fireEvent.mouseEnter(ex);
    fireEvent.mouseOver(li);

    expect(selectedList).toBe(null);
  });
});
