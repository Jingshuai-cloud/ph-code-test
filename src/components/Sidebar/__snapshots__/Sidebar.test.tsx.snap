// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Sidebar changes the class when hovered 1`] = `
<nav
  className="sc-bdfBwQ jPFBzQ"
>
  <ul
    onMouseLeave={[Function]}
  >
    <li
      className="sc-dlfnbm gYqwid"
    >
      <a
        className="sc-eCssSg iZhdA-D"
        href="/layer1"
      >
        AFFOGATO
      </a>
    </li>
    <li
      className="sc-dlfnbm gYqwid"
    >
      <span
        className="sc-hKgILt eCFKRY"
        onMouseEnter={[Function]}
      >
        ESPRESSO
      </span>
    </li>
  </ul>
</nav>
`;

exports[`Sidebar call the handleMenuSelection callback handler 1`] = `
<nav
  className="sc-bdfBwQ jPFBzQ"
>
  <ul
    onMouseLeave={[Function]}
  >
    <li
      className="sc-dlfnbm gYqwid"
    >
      <a
        className="sc-eCssSg iZhdA-D"
        href="/layer1"
      >
        AFFOGATO
      </a>
    </li>
    <li
      className="sc-dlfnbm gYqwid"
    >
      <span
        className="sc-hKgILt eCFKRY"
        onMouseEnter={[Function]}
      >
        ESPRESSO
      </span>
    </li>
  </ul>
</nav>
`;
